lib/eclipse/EclipseState/Aquancon.cpp

```diff
@@ -0,0 +1,159 @@
+/*
+  Copyright (C) 2017 TNO
+
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <opm/parser/eclipse/EclipseState/Aquancon.hpp>
+#include <algorithm>
+#include <iterator>
+
+namespace Opm {
+
+    Aquancon::Aquancon(const EclipseGrid& grid, const Deck& deck)
+    { 
+        if (!deck.hasKeyword("AQUANCON"))
+        {
+            getAquOutput();
+        }
+        
+        else {
+            const auto& aquanconKeyword = deck.getKeyword("AQUANCON");
+            // Resize the parameter vector container based on row entries in aquancon
+            m_aqurecord.resize(aquanconKeyword.size());
```

Comment:
I do not manage to follow everything in your implementation, but to me it does seem overly complex. Personally I would go through the records, expand into global indices and add directly to the `m_aquoutput` vector - completely bypassing the `m_aqurecord` variable, and also the quite complex book keeping in the private functions - just need to make sure the `(global_index, aquifer_id)` pair is not already in the output vector.

```C++
// Using small utility Box to assist with the calculation of indices:
Box grid_dims( grid.getNX(), grid.getNY(), grid.getNZ());

for (const auto& record :  aquanconKeyword) {
     int I1 = record.getitem("I1").get<int>(0);
     int I2 = record.getItem("I2").get<int>(0);
     ...
     int k2 = record.getItem("K2").get<int>(0);
     int aquifer_id = ..
     double influx_coeff = ...
     ....
     Box box(grid_dims, I1 - 1, I2 - 1, J1 - 1, J2 - 1, K1 - 1, K2 -1 );
    
     for (size_t g : box) {
            const auto * it = std::find_if( m_aquoutput.begin(), m_aquoutput.end(), [&g,&aquifer_id]( const struct AquanconOutput aq) { return (aq.aquiferID == aquifer_id && aq.global_index == g); } 
            if (it == m_aquauput.end())
               m_aquaoutput.emplace_back( { } )
     }
}
```

Reply:
The idea here is to output a vector which can identify all connections and essential quantities per aquifer instead of per AQUANCON record. m_aqurecord contains the per record information and m_aquoutput contains the per aquifer information. This is also what is asked for from the opm-simulators side. 

eg. The AQUANCON keyword could be as follows:


```
 AQUANCON
     1      1  33  1    33   1  2  J-  /
     1      1  33  1    33   1  2  J+  /
     1      1  33  1    33   1  2  I-  /
     2      1  33  1    33   1  2  J-  /
     1      1  33  1    33   1  2  I+  /
     2      1  33  1    33   1  2  J+  /
 /
```
 


Therefore, according to us eventhough the implementation is more complex, having this sort of a conversion to a per aquifer record has to be done somewhere, makes it easier to calculate face area fraction for each connection per aquifer and to implement logic such as: "Total number of grid blocks connected to aquifer must not exceed item 6 of AQUDIMS".

Reply:
> The idea here is to output a vector which can identify all connections and essential quantities per aquifer instead of per AQUANCON

OK - if per aquifer is most convenient downstream, then that is fine; although I do find that surprising. The records are purely input constructions and certainly relevant; what I (still ...) think would be the best was a datastructure for *one single aquifer connection* - i.e. *one flat vector* of something like:
```
struct AquiferCell {
     int aquifer_id;
     size_t global_index;
     double influx_coefficient;
     double flux_multiplier;
     face_enum face;
}
```
then it would be quite simple to perform ad-hoc calculations, filtering and so on based on this vector.

Reply:
With @kel85uk we discussed exactly what you are suggesting, but based on the requirement from simulators, the existing implementation there and for the reasons mentioned above, we decided the per aquifer output implementation was easier

Reply:
OK - but then I will stop nagging about that.